pkg/algorithm/hived_algorithm.go

```diff
@@ -666,13 +605,125 @@ func (h *HivedAlgorithm) updateVCDoomedBadCells(c CellChain, l CellLevel) {
+				internal.Key(pod), g.name, nodesNotInSuggested)
+			h.deletePreemptingAffinityGroup(g, pod)
+		} else {
+			groupPhysicalPlacement = g.physicalGpuPlacement
+			groupVirtualPlacement = g.virtualGpuPlacement
+			if preemptionVictims, _ = collectPreemptionVictims(groupPhysicalPlacement); len(preemptionVictims) == 0 {
+				klog.Infof("Preemption victims have been cleaned up for the preemptor affinity group %v", g.name)
+			}
+			g.preemptingPods[pod.UID] = pod
+		}
+	}
+	return groupPhysicalPlacement, groupVirtualPlacement, preemptionVictims, podIndex
+}
+
+// schedulePodFromNewGroup schedules a pod from a new affinity group, find placement for the group,
+// and checks if the group needs preemption.
+func (h *HivedAlgorithm) schedulePodFromNewGroup(
+	s *api.PodSchedulingSpec,
+	suggestedNodes common.Set,
+	phase internal.SchedulingPhase,
+	pod *core.Pod) (
+	groupPhysicalPlacement groupPhysicalPlacement,
+	groupVirtualPlacement groupVirtualPlacement,
+	preemptionVictims map[string]common.Set,
+	waitReason string) {
+
+	groupPhysicalPlacement, groupVirtualPlacement, nodesNotInSuggested := h.scheduleNewAffinityGroup(
+		pod, s, suggestedNodes)
+	if groupPhysicalPlacement == nil {
+		if nodesNotInSuggested.IsEmpty() {
+			waitReason = "insufficient capacity in physical cluster"
+			if CellPriority(s.Priority) >= minGuaranteedPriority {
+				waitReason = fmt.Sprintf("insufficient capacity in VC %v", s.VirtualCluster)
+			}
+		} else {
+			// for an unallocated group,
+			// we will keep it waiting if not all of its pods are scheduled to suggested nodes
+			waitReason = fmt.Sprintf("affinity group is decided to be scheduled to some nodes "+
+				"not within K8s suggested nodes: %v", nodesNotInSuggested)
+		}
+		return nil, nil, nil, waitReason
+	}
+	preemptionVictims, overlappingPreemptors := collectPreemptionVictims(groupPhysicalPlacement)
+	// we allow a new preemption only when in Preempting phase
+	// and the placement is fully within suggested nodes
+	if phase == internal.PreemptingPhase {
+		for preemptor := range overlappingPreemptors.Items() {
+			klog.Infof("[%v]: Canceling affinity group %v's preemption because it is "+
+				"further preempted by a higher-priority affinity group %v",
+				internal.Key(pod), preemptor.(*AlgoAffinityGroup).name, s.AffinityGroup.Name)
+			h.deletePreemptingAffinityGroup(preemptor.(*AlgoAffinityGroup), pod)
+		}
+		if len(preemptionVictims) != 0 {
+			h.createPreemptingAffinityGroup(s, groupPhysicalPlacement, groupVirtualPlacement, pod)
+		}
+	} else if len(preemptionVictims) != 0 {
+		klog.Infof("[%v]: Found preemption victims %v, but we do not allow preemption in Filtering "+
+			"phase because K8s won't call preempt, creating preemption state here is misleading",
+			internal.Key(pod), victimsToString(preemptionVictims))
+		groupPhysicalPlacement = nil
+		groupVirtualPlacement = nil
+		preemptionVictims = nil
+		waitReason = "affinity group wants to preempt in Filtering phase"
+	}
```

Comment:
Can we still return PodPreemptInfo in this case, but just do not need to createPreemptingAffinityGroup for it?

Because this is to  align:
```
// 1. If the SchedulerAlgorithm found sufficient free resource, only PodBindInfo
//    should be set.
//    If the SchedulerAlgorithm found sufficient preemptible resource, only
//    PodPreemptInfo should be set.
//    Otherwise, only PodWaitInfo can be optionally set.
```

And outside waitReason:

```
  klog.Infof(logPfx+
   "Pod is waiting for preemptRoutine as preemptible resource appeared: %v",
   common.ToJson(failedNodes))

  waitReason := "Pod is waiting for preemptible or free resource to appear"
```